File: truth/frontend/src/requests/schemas.ts
/**
 * Every call around the schemas - the declarations that decide what attributes an assumption carries.
 */

import type { SchemaDetail, SchemaDraft, SchemaRevisionDraft, SchemaSummary } from '@/models/schema'
import { client } from '@/requests/client'
import { readAll } from '@/requests/paging'

const SCHEMAS_PATH = '/schema'

/**
 * Read one window of the schemas.
 */
const listSchemas = async (offset: number, limit: number): Promise<SchemaSummary[]> => {
  const response = await client.get<SchemaSummary[]>(SCHEMAS_PATH, { params: { offset, limit } })

  return response.data
}

/**
 * Read every declared schema, as the listing hands them over.
 */
const readAllSchemas = async (onProgress?: (loaded: number) => void): Promise<SchemaSummary[]> =>
  readAll(listSchemas, onProgress)

/**
 * Read one schema whole, at the revision it is currently at.
 */
const readSchema = async (schemaId: string): Promise<SchemaDetail> => {
  const response = await client.get<SchemaDetail>(`${SCHEMAS_PATH}/${schemaId}`)

  return response.data
}

/**
 * Read the latest revision of one schema, which is what a form is built from.
 */
const readLatestSchema = async (schemaId: string): Promise<SchemaDetail> => {
  const response = await client.get<SchemaDetail>(`${SCHEMAS_PATH}/${schemaId}/latest`)

  return response.data
}

/**
 * Read one schema addressed by its name rather than by its identifier.
 */
const readSchemaByName = async (name: string): Promise<SchemaDetail> => {
  const response = await client.get<SchemaDetail>(`${SCHEMAS_PATH}/name/${encodeURIComponent(name)}`)

  return response.data
}

/**
 * Declare a new schema and hand back the identifier it was given.
 */
const createSchema = async (draft: SchemaDraft): Promise<string> => {
  const response = await client.post<{ id: string }>(SCHEMAS_PATH, draft)

  return response.data.id
}

/**
 * Revise a schema addressed by its identifier, which is the only way a declaration is ever changed.
 */
const createSchemaRevision = async (schemaId: string, draft: SchemaRevisionDraft): Promise<string> => {
  const response = await client.post<{ id: string }>(`${SCHEMAS_PATH}/${schemaId}/revisions`, draft)

  return response.data.id
}

/**
 * Revise a schema addressed by its name.
 */
const createSchemaRevisionByName = async (name: string, draft: SchemaRevisionDraft): Promise<string> => {
  const response = await client.post<{ id: string }>(
    `${SCHEMAS_PATH}/name/${encodeURIComponent(name)}/revisions`,
    draft,
  )

  return response.data.id
}

export {
  createSchema,
  createSchemaRevision,
  createSchemaRevisionByName,
  listSchemas,
  readAllSchemas,
  readLatestSchema,
  readSchema,
  readSchemaByName,
}

File: truth/backend/src/truth_service/api/schemas.py
"""
The endpoints of the schemas, which declare what attributes an assumption carries.

:date: 2026-09-08
"""
# ----- IMPORTS ----- #

from fastapi import APIRouter, status

from truth_service.api.pagination import LIMIT_QUERY, OFFSET_QUERY
from truth_service.dependencies import SchemaServiceDependency
from truth_service.models.schema import (
    SchemaCreateRequest,
    SchemaDetailResponse,
    SchemaRevisionRequest,
    SchemaSummaryResponse,
)

# ----- CONSTS ----- #

ROUTER: APIRouter = APIRouter(prefix="/schema", tags=["schemas"])

# ----- FUNCTIONS ----- #


@ROUTER.get("", response_model=list[SchemaSummaryResponse])
async def list_schemas(
    service: SchemaServiceDependency,
    offset: int = OFFSET_QUERY,
    limit: int = LIMIT_QUERY,
) -> list[SchemaSummaryResponse]:
    """
    Read one window of the declarations, each at the revision it currently stands at.

    :param service: Owner of the declarations.
    :param offset: Amount of declarations skipped before collecting.
    :param limit: Largest amount of declarations that is returned.
    :return: The declarations of the window.
    """
    return await service.list_schemas(offset=offset, limit=limit)


@ROUTER.post("", response_model=SchemaDetailResponse, status_code=status.HTTP_201_CREATED)
async def create_schema(request: SchemaCreateRequest, service: SchemaServiceDependency) -> SchemaDetailResponse:
    """
    Declare a new schema, which is the first revision of a new lineage.

    :param request: Declaration supplied by the caller.
    :param service: Owner of the declarations.
    :return: The declaration as it was stored, carrying the identifier it was given.
    """
    return await service.create_schema(request=request)


@ROUTER.get("/name/{name}", response_model=SchemaDetailResponse)
async def read_schema_by_name(name: str, service: SchemaServiceDependency) -> SchemaDetailResponse:
    """
    Read the current revision of the declaration answering to one name.

    :param name: Name of the declaration.
    :param service: Owner of the declarations.
    :return: The current revision, whole.
    """
    return await service.get_by_name(name=name)


@ROUTER.post("/name/{name}/revisions", response_model=SchemaDetailResponse, status_code=status.HTTP_201_CREATED)
async def revise_schema_by_name(
    name: str,
    request: SchemaRevisionRequest,
    service: SchemaServiceDependency,
) -> SchemaDetailResponse:
    """
    Write a new revision of the declaration answering to one name.

    :param name: Name of the declaration.
    :param request: The declaration as it now stands, and why it was changed.
    :param service: Owner of the declarations.
    :return: The revision that was written.
    """
    return await service.revise_schema(key=name, request=request)


@ROUTER.get("/{schema_id}/latest", response_model=SchemaDetailResponse)
async def read_latest_schema(schema_id: str, service: SchemaServiceDependency) -> SchemaDetailResponse:
    """
    Read the current revision of a declaration addressed by any of its identifiers.

    :param schema_id: Identifier of a revision or of the lineage.
    :param service: Owner of the declarations.
    :return: The current revision, whole.
    """
    return await service.get_latest(key=schema_id)


@ROUTER.post("/{schema_id}/revisions", response_model=SchemaDetailResponse, status_code=status.HTTP_201_CREATED)
async def revise_schema(
    schema_id: str,
    request: SchemaRevisionRequest,
    service: SchemaServiceDependency,
) -> SchemaDetailResponse:
    """
    Write a new revision of a declaration, which is the only way a declaration is ever changed.

    :param schema_id: Identifier of a revision or of the lineage.
    :param request: The declaration as it now stands, and why it was changed.
    :param service: Owner of the declarations.
    :return: The revision that was written.
    """
    return await service.revise_schema(key=schema_id, request=request)


@ROUTER.get("/{schema_id}", response_model=SchemaDetailResponse)
async def read_schema(schema_id: str, service: SchemaServiceDependency) -> SchemaDetailResponse:
    """
    Read exactly the revision that was addressed, falling back to the current one of its lineage.

    :param schema_id: Identifier of a revision, of the lineage, or the name of the declaration.
    :param service: Owner of the declarations.
    :return: That revision, whole.
    """
    return await service.get_revision(identifier=schema_id)
